substrate/network-libp2p/src/service.rs

```diff
@@ -0,0 +1,1106 @@
+
+		let local_peer_id = network_state.local_public_key().clone().into_peer_id();
+		// TODO: debug! instead?
+		info!(target: "sub-libp2p", "Local node id = {:?}", local_peer_id);
+
+		let kad_system = KadSystem::without_init(KadSystemConfig {
+			parallelism: 3,
+			local_peer_id: local_peer_id.clone(),
+			kbuckets_timeout: Duration::from_secs(10),
+			request_timeout: Duration::from_secs(10),
+			known_initial_peers: network_state.known_peers().collect(),
+		});
+
+		let shared = Arc::new(Shared {
+			network_state,
+			protocols: RwLock::new(Default::default()),
+			kad_system,
+			kad_upgrade: KadConnecConfig::new(),
+			config,
+			timeouts_register_tx: RwLock::new(mpsc::unbounded().0),
+			original_listened_addr: RwLock::new(None),
+			listened_addrs: RwLock::new(Vec::new()),
+		});
+
+		Ok(NetworkService {
+			shared,
+			bg_thread: Mutex::new(None),
+		})
+	}
+
+	/// Register a new protocol handler with the event loop.
+	pub fn register_protocol(
+		&self,
+		handler: Arc<NetworkProtocolHandler + Send + Sync>,
+		protocol: ProtocolId,
+		versions: &[(u8, u8)]
+	) {
+		if self.shared.network_state.has_connected_peer() {
+			// TODO: figure out if that's correct
+			warn!(target: "sub-libp2p", "a new network protocol was registered \
+				while the service was already active ; this is a programmer \
+				error");
+		}
+
+		self.shared.protocols.write().0
+			.push(RegisteredProtocol::new(handler.clone(), protocol, versions));
+
+		handler.initialize(&NetworkContextImpl {
+			inner: self.shared.clone(),
+			protocol: protocol.clone(),
+			current_peer: None,
+		});
+	}
+
+	/// Returns network configuration.
+	pub fn config(&self) -> &NetworkConfiguration {
+		&self.shared.config
+	}
+
+	pub fn external_url(&self) -> Option<String> {
+		// TODO: in the context of libp2p, it is hard to define what an external
+		// URL is, as different nodes can have multiple different ways to
+		// reach us
+		self.shared.original_listened_addr.read().as_ref()
+			.map(|addr|
+				format!("{}/p2p/{}", addr,
+					self.shared.kad_system.local_peer_id().to_base58())
+			)
+	}
+
+	/// Start network IO
+	// TODO (design): the notion of having a `NetworkService` alive should mean
+	// that it is running ; the `start` and `stop` functions are bad design
+	pub fn start(&self) -> Result<(), (Error, Option<SocketAddr>)> {
+		// TODO: check that service is started already?
+
+		*self.shared.protocols.write() = Default::default();
+
+		// Channel we use to signal success or failure of the bg thread
+		// initialization process.
+		let (init_tx, init_rx) = sync_mpsc::channel();
+		// Channel the main thread uses to signal the bg thread that it
+		// should stop
+		let (close_tx, close_rx) = oneshot::channel();
+		let (timeouts_register_tx, timeouts_register_rx) = mpsc::unbounded();
+		let shared = self.shared.clone();
+		let join_handle = thread::spawn(move || {
+			// Tokio core that is going to run everything in this thread.
+			let mut core = match Core::new() {
```

Comment:
could this be easily replaced with a `tokio::current_thread::Runtime`?

Reply:
Libp2p-rs uses the old tokio-core thing, because I admit I'm a bit lost in tokio versions, so no.

Reply:
are there plans to upgrade it? we're mostly using new-tokio in polkadot.

Reply:
I'm not sure that the new `tokio v0.1` thing can compile to emscripten, may have to check if it works before upgrading.